src/diagonal.works/b6/ui/lines.go

```diff
@@ -0,0 +1,707 @@
+}
+
+func fillSubstackFromExpression(lines *pb.SubstackProto, expression *pb.NodeProto, root bool) {
+	if call, ok := expression.Node.(*pb.NodeProto_Call); ok {
+		if call.Call.Pipelined {
+			left := call.Call.Args[0]
+			right := &pb.NodeProto{
+				Node: &pb.NodeProto_Call{
+					Call: &pb.CallNodeProto{
+						Function: call.Call.Function,
+						Args:     call.Call.Args[1:],
+					},
+				},
+			}
+			fillSubstackFromExpression(lines, left, false)
+			fillSubstackFromExpression(lines, right, false)
+			return
+		}
+	}
+	_, isLiteral := expression.Node.(*pb.NodeProto_Literal)
+	if expression, ok := api.UnparseNode(expression); ok {
+		if !isLiteral || !root {
+			lines.Lines = append(lines.Lines, &pb.LineProto{
+				Line: &pb.LineProto_Expression{
+					Expression: &pb.ExpressionLineProto{
+						Expression: expression,
+					},
+				},
+			})
+		}
+	} else {
+		lines.Lines = append(lines.Lines, &pb.LineProto{
+			Line: &pb.LineProto_Error{
+				Error: &pb.ErrorLineProto{
+					Error: "can't convert function",
+				},
+			},
+		})
+	}
+}
+
+func fillSubstackFromCollection(substack *pb.SubstackProto, c api.Collection, response *pb.UIResponseProto) error {
+	// TODO: Set collection title based on collection contents
+	if countable, ok := c.(api.Countable); ok {
+		substack.Lines = append(substack.Lines, &pb.LineProto{
+			Line: &pb.LineProto_ValuePair{
+				ValuePair: &pb.ValuePairLineProto{
+					First:  atomFromString("Collection"),
+					Second: atomFromString(strconv.Itoa(countable.Count())),
+				},
+			},
+		})
+	} else {
+		substack.Lines = append(substack.Lines, &pb.LineProto{
+			Line: &pb.LineProto_Value{
+				Value: &pb.ValueLineProto{
+					Atom: atomFromString("Collection"),
+				},
+			},
+		})
+	}
+
+	keys := make([]interface{}, 0, 8)
+	values := make([]interface{}, 0, 8)
+	var err error
+	keys, values, err = fillKeyValues(c, keys, values)
+	if err != nil {
+		return err
+	}
+	if isFeatureCollection(keys, values) || isArrayCollection(keys, values) {
+		for i := range values {
+			if i < CollectionLineLimit {
+				line := valueLineFromValue(values[i])
+				substack.Lines = append(substack.Lines, line)
+			} else {
+				break
+			}
+		}
+	} else {
+		for i := range keys {
+			if i < CollectionLineLimit {
+				line := valuePairLineFromValues(keys[i], values[i])
+				substack.Lines = append(substack.Lines, line)
+			} else {
+				break
+			}
+		}
+	}
+
+	for i := range keys {
+		if id, ok := keys[i].(b6.Identifiable); ok {
+			highlightInResponse(response, id.FeatureID())
+		}
+		if id, ok := values[i].(b6.Identifiable); ok {
+			highlightInResponse(response, id.FeatureID())
+		}
+	}
+	return nil
+}
+
+func fillSubstackFromHistogram(substack *pb.SubstackProto, c *api.HistogramCollection) error {
+	keys, values, err := fillKeyValues(c, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	total := 0
+	begin := len(substack.Lines)
+	for i, key := range keys {
+		// TODO: Factor out w/ collection-key-or-value fill when we rework blocks
```

Comment:
remove todo

Reply:
I'm going to do that in a follow CL, if that's OK.